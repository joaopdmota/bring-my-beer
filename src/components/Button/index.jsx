// @flow
import React, { type Node } from 'react'
import styled, { css } from 'styled-components'
import { lighten } from 'polished'

const Button = styled.div`
  user-select: none;
  border: 0px;
  border-radius: ${props => props.theme.buttonRadius};
  background-color: ${props =>
    !props.disabled
      ? props.theme[props.appearance]
      : lighten(0.3, props.theme[props.appearance])};
  color: ${props => props.theme.txt};
  padding: 12px;
  font-size: 12px;
  cursor: ${props => (!props.disabled ? 'pointer' : 'default')};
  outline: none;
  &:focus,
  &:active {
    background-color: ${props =>
      !props.disabled
        ? props.theme[props.appearance]
        : lighten(0.3, props.theme[props.appearance])};
  }
  &:focus {
    outline: none;
    box-shadow: 0px 1px 1px rgba(0, 0, 0, 0.5);
  }
  ${props =>
    !props.disabled &&
    css`
      &:hover {
        background-color: ${props =>
          lighten(0.2, props.theme[props.appearance])};
      }
    `}
`

type ButtonProps = {
  children?: Node,
  onClick?: void => void,
  className?: string,
  appearance?: string,
  disabled?: Boolean,
}

export default ({
  children,
  onClick,
  className,
  appearance = 'mainColor',
  disabled = false,
}: ButtonProps) => {
  return (
    <Button
      className={className}
      onClick={e => {
        if (!disabled && onClick) {
          onClick(e)
        } else return
      }}
      appearance={appearance}
      role="button"
      disabled={disabled}
    >
      <div>{children}</div>
    </Button>
  )
}
