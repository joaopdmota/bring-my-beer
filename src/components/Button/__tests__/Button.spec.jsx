import React from 'react'
import Button from '../index'
import { shallow } from 'enzyme'

describe('Button component', () => {
  it('render default button', () => {
    const button = <Button>Foo bar</Button>
    expect(button).toMatchSnapshot()
  })
  it('render success button', () => {
    const button = <Button appearance="success">Foo bar</Button>
    expect(button).toMatchSnapshot()
  })
  it('render error button', () => {
    const button = <Button appearance="success">Foo bar</Button>
    expect(button).toMatchSnapshot()
  })
  it('render button with children', () => {
    const child = <div>Foo bar</div>
    const button = shallow(<Button appearance="success">{child}</Button>)
    expect(button.find(child)).toBeTruthy()
  })
  it('render a button with onClick function', () => {
    const fnClick = jest.fn()
    const button = shallow(<Button onClick={fnClick}>Foo Bar</Button>)

    button.simulate('click')
    expect(fnClick).toHaveBeenCalled()
  })
  it('render a disabled button', () => {
    const fnClick = jest.fn()
    const button = shallow(
      <Button disabled onClick={fnClick}>
        Foo Bar
      </Button>,
    )

    button.simulate('click')
    expect(fnClick).not.toHaveBeenCalled()
  })
})
