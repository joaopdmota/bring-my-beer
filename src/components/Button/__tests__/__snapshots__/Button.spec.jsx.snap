// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component render default button 1`] = `
<_default>
  Foo bar
</_default>
`;

exports[`Button component render error button 1`] = `
<_default
  appearance="success"
>
  Foo bar
</_default>
`;

exports[`Button component render success button 1`] = `
<_default
  appearance="success"
>
  Foo bar
</_default>
`;
